kernel/kernel-default/src/main/java/io/delta/kernel/parquet/RowConverter.java

```diff
@@ -0,0 +1,153 @@
+package io.delta.kernel.parquet;
+
+import static io.delta.kernel.DefaultKernelUtils.findFieldType;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import static java.util.Objects.requireNonNull;
+import java.util.Optional;
+import org.apache.parquet.io.api.Converter;
+import org.apache.parquet.io.api.GroupConverter;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.Type;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.data.DefaultColumnarBatch;
+import io.delta.kernel.data.vector.DefaultStructVector;
+import io.delta.kernel.types.DataType;
+import io.delta.kernel.types.StructField;
+import io.delta.kernel.types.StructType;
+
+class RowConverter
+    extends GroupConverter
+    implements ParquetConverters.BaseConverter
+{
+    private final StructType readSchema;
+    private final Converter[] converters;
+    // The delta may request columns that don't exists in Parquet
+    // This map is to track the ordinal known to Parquet reader to the converter array ordinal.
+    // If a column is missing, a dummy converter added to the `converters` array and which
+    // generates all null vector at the end.
+    private final Map<Integer, Integer> parquetOrdinalToConverterOrdinal;
+
+    // Working state
+    private int currentRowIndex;
+    private boolean[] nullability;
+
+    RowConverter(
+        int maxBatchSize,
+        StructType readSchema,
+        GroupType fileSchema)
+    {
+        this.readSchema = requireNonNull(readSchema, "readSchema is not null");
+        List<StructField> fields = readSchema.fields();
+        this.converters = new Converter[fields.size()];
+        this.parquetOrdinalToConverterOrdinal = new HashMap<>();
+
+        // Initialize the working state
+        this.nullability = ParquetConverters.initNullabilityVector(maxBatchSize);
+
+        int parquetOrdinal = 0;
```

Comment:
We can't work without that assumption, because the reader client expects the data in the same order as the readSchema. If it is better, I can move the pruneSchema utility method under ParquetBatchReader. 

This code needs to revisited once we starts supporting column mapping id based mode.

`GroupType::getFieldIndex` not available for primitive types.

Reply:
I don't really understand why we can't just change the mapping we store in `parquetOrdinalToConverterOrdinal`? That allows us to keep the returned data in the order of the `readSchema`.

Reply:
And in the future we can update how we map between the two to accomodate for id based mapping.

Reply:
Can I address this separately? I am not fully convinced. Noted in the list.